pkg/cache/simple.go

```diff
@@ -46,119 +49,169 @@ type SimpleCache struct {
 	mu sync.Mutex
 }
 
-// MakeSimpleCache initializes a simple cache.
+// Snapshot is an internally consistent snapshot of xDS resources.
+// Snapshots should have distinct versions per node group.
+type Snapshot struct {
+	version   string
+	resources map[ResponseType][]proto.Message
+}
+
+// NewSnapshot creates a snapshot from response types and a version.
+func NewSnapshot(version string,
+	endpoints []proto.Message,
+	clusters []proto.Message,
+	routes []proto.Message,
+	listeners []proto.Message) Snapshot {
+	return Snapshot{
+		version: version,
+		resources: map[ResponseType][]proto.Message{
+			EndpointResponse: endpoints,
+			ClusterResponse:  clusters,
+			RouteResponse:    routes,
+			ListenerResponse: listeners,
+		},
+	}
+}
+
+// NewSimpleCache initializes a simple cache.
 // callback function is called on every new cache key and response type if there is no response available.
-// callback is executed in a go-routine.
-func MakeSimpleCache(groups NodeGroup, callback func(Key, ResponseType)) Cache {
+// callback is executed in a go-routine and can be called multiple times prior to receiving a snapshot.
+func NewSimpleCache(groups NodeGroup, callback func(Key)) Cache {
 	return &SimpleCache{
-		responses:  make(map[Key]map[ResponseType][]proto.Message),
-		versions:   make(map[Key]map[ResponseType]int),
-		watches:    make(map[Key]map[ResponseType]map[int64]Watch),
+		snapshots:  make(map[Key]Snapshot),
+		watches:    make(map[Key]map[int64]Watch),
 		callback:   callback,
 		watchCount: 0,
 		groups:     groups,
 	}
 }
 
-func (cache *SimpleCache) SetResource(group Key, typ ResponseType, resources []proto.Message) {
+// SetSnapshot updates the simple cache with a snapshot for a node group.
+func (cache *SimpleCache) SetSnapshot(group Key, snapshot Snapshot) error {
+	// TODO(kuat) validate snapshot for types and internal consistency
+
 	cache.mu.Lock()
 	defer cache.mu.Unlock()
 
 	// update the existing entry
-	if _, ok := cache.responses[group]; !ok {
-		cache.responses[group] = make(map[ResponseType][]proto.Message)
-		cache.versions[group] = make(map[ResponseType]int)
+	cache.snapshots[group] = snapshot
+
+	// trigger existing watches
+	if _, ok := cache.watches[group]; ok {
+		for _, watch := range cache.watches[group] {
+			respond(watch, snapshot, group)
+		}
+		// discard all watches; the client must request a new watch to receive updates and ACK/NACK
+		delete(cache.watches, group)
 	}
 
-	cache.responses[group][typ] = resources
+	return nil
+}
 
-	// increment version
-	cache.versions[group][typ]++
+// Respond to a watch with the snapshot value. The value channel should have capacity not to block.
+// TODO(kuat) this responds immediately and can cause the remote node to spin if it consistently fails to ACK the update
+func respond(watch Watch, snapshot Snapshot, group Key) {
+	typ := watch.Type
```

Comment:
nit: I'm a fan of grouping declaration/assignments, but not sure if this is recommended. cc: @davecheney 

```go
var (
    typ        = watch.Type
    resources  = snapshot.resources[typ]
    version    = snapshot.version
)
```

Reply:
let me know what the stylistic recommendation is :)